src/server/requestHandlers.js

```diff
@@ -222,42 +188,85 @@ sjrk.storyTelling.server.saveStoryToDatabase = function (dataSource, binaryRenam
+ * @param {Boolean} authoringEnabled - a server-level flag to indicate whether authoring is enabled
  */
-sjrk.storyTelling.server.buildBinaryRenameMap = function (blocks, files) {
-    // key-value pairs of original filename : generated filename
-    // this is used primarily by tests, but may be of use
-    // to client-side components too
-    var binaryRenameMap = {};
-
-    fluid.each(blocks, function (block) {
-        if (block.blockType === "image" || block.blockType === "audio" || block.blockType === "video") {
-            if (block.fileDetails) {
-                // Look for the uploaded file matching this block
-                var mediaFile = fluid.find_if(files, function (file) {
-                    return file.originalname === block.fileDetails.name;
-                });
+sjrk.storyTelling.server.handleSaveStoryFile = function (request, dataSource, authoringEnabled) {
+    if (!authoringEnabled) {
+        request.events.onError.fire({
+            isError: true,
+            message: "Saving is currently disabled."
+        });
+
+        return;
+    }
 
-                // If we find a match, update the media URL. If not, clear it.
-                if (mediaFile) {
-                    sjrk.storyTelling.server.setMediaBlockUrl(block, mediaFile.filename);
-                    binaryRenameMap[mediaFile.originalname] = mediaFile.filename;
-                } else {
-                    sjrk.storyTelling.server.setMediaBlockUrl(block, null);
-                }
+    var id = request.req.params.id;
+
+    // verify that the story exists and isn't published before continuing
+    dataSource.get({directStoryId: id}).then(function (story) {
+        if (story && !story.published && request.req.file) {
+            // if the file is an image, rotate it based on its EXIF data
+            if (request.req.file.mimetype &&
+                request.req.file.mimetype.indexOf("image") === 0) {
+
+                sjrk.storyTelling.server.rotateImageFromExif(request.req.file).then(function () {
+                    request.events.onSuccess.fire(request.req.file.destination + "/" + request.req.file.filename);
+                }, function (error) {
+                    request.events.onError.fire({
+                        errorCode: error.errorCode,
+                        isError: true,
+                        message: error.message || "Unknown error in image rotation."
+                    });
+                });
             } else {
-                sjrk.storyTelling.server.setMediaBlockUrl(block, null);
+                request.events.onSuccess.fire(request.req.file.destination + "/" + request.req.file.filename);
             }
+        } else {
+            request.events.onError.fire({
+                isError: true,
+                message: "Error saving file: file was not provided or story is already published"
+            });
         }
+    }, function (err) {
+        request.events.onError.fire({
```

Comment:
Are we returning proper [http response codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) with all of the errors?

Reply:
We aren't, currently. I've been following the example set by `fluid.promise` and returning `isError: true` and a `message` string with every error, but HTTP status codes would be good to add as well. Would you like this as part of this work, or should we file a Jira to tackle it project-wide?

Reply:
Let's file this as a new JIRA and review things as a whole.